src/node_sets.cpp

```diff
@@ -0,0 +1,317 @@
+                }
+            } else if (array[0].is_string()) {
+                if (attribute == "node_id") {
+                    throw SonataError("'node_id' must be numeric or a list of numbers");
+                }
+
+                std::vector<std::string> values;
+                for (auto& inner_el : array.items()) {
+                    values.emplace_back(inner_el.value().get<std::string>());
+                }
+
+                ret.emplace_back(new NodeSetBasicRule<std::string>(attribute, std::move(values)));
+            } else {
+                throw SonataError("Unknown array type");
+            }
+        }
+    }
+    return ret;
+}
+
+void parse_basic(const json& j, std::map<std::string, NodeSetRules>& node_sets) {
+    for (auto& el : j.items()) {
+        if (el.value().is_object()) {
+            node_sets[el.key()] = _dispatch_node(el.value());
+        }
+    }
+}
+
+void check_compound(const std::map<std::string, NodeSetRules>& node_sets,
+                    const std::map<std::string, CompoundTargets>& compound_rules,
+                    size_t depth) {
+    if (depth > MAX_COMPOUND_RECURSION) {
+        throw SonataError("Compound node_set recursion depth exceeded");
+    }
+
+    for (auto const& it : compound_rules) {
+        for (auto const& name : it.second) {
+            if (compound_rules.count(name) > 0) {
+                check_compound(node_sets, compound_rules, depth + 1);
+            } else if (node_sets.count(name) == 0) {
+                throw SonataError(fmt::format("Missing '{}' from node_sets", name));
+            }
+        }
+    }
+}
+
+void parse_compound(const json& j, std::map<std::string, NodeSetRules>& node_sets) {
+    std::map<std::string, CompoundTargets> compound_rules;
+    for (auto& el : j.items()) {
+        if (el.value().is_array()) {
+            CompoundTargets targets;
+            for (const auto& name : el.value()) {
+                if (!name.is_string()) {
+                    throw SonataError("All compound node types must be strings");
+                }
+
+                targets.emplace_back(name);
+            }
+            compound_rules[el.key()] = targets;
+        }
+    }
+
+    check_compound(node_sets, compound_rules, 0);
+
+    for (const auto& rule : compound_rules) {
+        NodeSetRules rules;
+        rules.emplace_back(new NodeSetCompoundRule(rule.first, rule.second));
+        node_sets.emplace(rule.first, std::move(rules));
+    }
+}
+}  // anonymous namespace
+
+NodeSets::NodeSets(const std::string& content) {
+    json j = json::parse(content);
+    if (!j.is_object()) {
+        throw SonataError("Top level node_set must be an object");
+    }
+
+    // Need to two pass parsing the json so that compound lookup can rely
+    // on all the basic rules existing
+    parse_basic(j, node_sets_);
+    parse_compound(j, node_sets_);
+}
+
+Selection NodeSets::materialize(const std::string& name, const NodePopulation& population) const {
```

Comment:
maybe "apply" is a better name, no ?

Reply:
I'm agnostic, we'll see what others say.

Reply:
I vote for `summon`.